### PR TITLE
KIP-140: ACL operations

### PR DESCRIPTION
**Pull request by emasab**
implements bindings for ACL operations:

CreateAcls
DescribeAcls
DeleteAcls

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -7,16 +7,14 @@ env:
 matrix:
   include:
   - os: linux
-    dist: trusty # Ubuntu 14.04
-    dotnet: 2.1.300
+    dist: focal # Ubuntu 20.04
+    dotnet: 5.0.408
```
Comment (mhowlett) [inline .travis.yml:11]: cool, thanks. since we're moving away from travis, i was just letting it rot ..

File: examples/AdminClient/Program.cs [hunk 2]
```diff
@@ -19,7 +19,8 @@
 using System;
 using System.Threading.Tasks;
 using Confluent.Kafka.Admin;
-
+using System.Linq;
+using System.Collections.Generic;
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:24]: convention in this repo is two lines between using statements and namespace.

File: examples/AdminClient/Program.cs [hunk 3]
```diff
@@ -84,7 +85,202 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
                 }
                 catch (CreateTopicsException e)
                 {
-                    Console.WriteLine($"An error occured creating topic {e.Results[0].Topic}: {e.Results[0].Error.Reason}");
+                    Console.WriteLine($"An error occurred creating topic {e.Results[0].Topic}: {e.Results[0].Error.Reason}");
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:88]: hey, i thought i was the native english speaker 😉

File: examples/AdminClient/Program.cs [hunk 4]
```diff
@@ -93,10 +289,11 @@ public static async Task Main(string[] args)
         {
             if (args.Length < 2)
             {
-                Console.WriteLine("usage: .. <bootstrapServers> <list-groups|metadata|library-version|create-topic> [topic-name]");
+                Console.WriteLine("usage: .. <bootstrapServers> <list-groups|metadata|library-version|create-topic|create-acls|describe-acls|delete-acls> ..");
                 System.Environment.Exit(1);
             }
-            
+
+            var commandArgs = args.Skip(2).ToArray();
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:296]: maybe introduce a variable for bootstrapServers here, and don't use args[0] below.

File: examples/AdminClient/Program.cs [hunk 5]
```diff
@@ -112,6 +309,15 @@ public static async Task Main(string[] args)
                 case "create-topic":
                     await CreateTopicAsync(args[0], args[2]);
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:310]: maybe change this to pass commandArgs for consistency with the below (and it's nicer).

File: examples/AdminClient/Program.cs [hunk 6]
```diff
@@ -84,7 +85,202 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                    Console.WriteLine($"An error occurred creating topic {e.Results[0].Topic}: {e.Results[0].Error.Reason}");
+                }
+            }
+        }
+
+        static List<T> ParseAclBindings<T>(string[] args) where T: AclBinding, new()
+        {
+            int nAclBindings = args.Length / 7;
+            var aclBindings = new List<T>();
+            for (int i = 0; i < nAclBindings; i++)
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:97]: nit: i use ++i in this codebase, optionally feel free to change.

File: src/Confluent.Kafka/Admin/AclBindingFilter.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+
+namespace Confluent.Kafka.Admin
+{
+    /// <summary>
+    /// Represents a filter used to return a list of ACL bindings matching some or all of its attributes.
```
Comment (mhowlett) [inline src/Confluent.Kafka/Admin/AclBindingFilter.cs:21]: the convention for doc strings in this repo is to tab-indent. have a look elsewhere for style.

File: src/Confluent.Kafka/Admin/AclBinding.cs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        /// <summary>
+        ///     The resource type.
+        /// </summary>
+        public ResourceType Type { get; set; }
+
+        /// <summary>
+        /// The resource name, which depends on the resource type.
+        /// For ResourceBroker the resource name is the broker id.
+        /// </summary>
+	    public string Name { get; set; }
```
Comment (mhowlett) [inline src/Confluent.Kafka/Admin/AclBinding.cs:39]: indentation

File: examples/AdminClient/Program.cs [hunk 9]
```diff
@@ -84,7 +85,202 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
                 }
                 catch (CreateTopicsException e)
                 {
-                    Console.WriteLine($"An error occured creating topic {e.Results[0].Topic}: {e.Results[0].Error.Reason}");
+                    Console.WriteLine($"An error occurred creating topic {e.Results[0].Topic}: {e.Results[0].Error.Reason}");
+                }
+            }
+        }
+
+        static List<T> ParseAclBindings<T>(string[] args) where T: AclBinding, new()
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:93]: although the inheritance relationship may be convenient, this comes across as a bit weird to me. see comment elsewhere. i notice the Java implementation has a toFilter method on the AclBinding class, so probably do similar instead and make use of that i think.

File: examples/AdminClient/Program.cs (hunk 3 above)
Comment (emasab) [reply examples/AdminClient/Program.cs:88]: It's the automatic spell check I've installed to avoid my errors :smile:

File: src/Confluent.Kafka/Admin/AclBinding.cs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+        /// <summary>
+        /// The resource pattern, relative to the name.
+        /// </summary>
+        public ResourcePatternType ResourcePatternType { get; set; }
+
+        /// <summary>
+        /// The principal this AclBinding refers to.
+        /// </summary>
+        public string Principal { get; set; }
```
[the remark this answers is not in the document]
Comment (emasab) [reply src/Confluent.Kafka/Admin/AclBinding.cs:49]: In librdkafka and the other non-Java clients it's a flat structure, but if you want it more similar to the Java one it's possible.

**Comment by emasab** [comment]
Thanks @mhowlett ! I've addressed the current comments

File: examples/AdminClient/Program.cs [hunk 11]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+            foreach (AclBinding aclBinding in aclBindings)
+            {
+                Console.WriteLine($"\t{aclBinding}");
+            }
+        }
+
+        static async Task CreateAclsAsync(string bootstrapServers, string[] commandArgs)
+        {
+            var numCommandArgs = commandArgs.Length;
+            var printUsage = numCommandArgs == 0 || numCommandArgs % 7 != 0;
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:165]: this logic better belongs in ParseAclBindings, and throw an exception if there's an issue. then handle that here to print usage.

File: examples/AdminClient/Program.cs [hunk 12]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+            List<AclBinding> aclBindings = null;
+            if (!printUsage)
+            {
+                aclBindings = ParseAclBindings(commandArgs);
+            }
+            printUsage = aclBindings == null;
+
+            if (printUsage)
+            {
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:174]: this block becomes the catch {} block.

File: examples/AdminClient/Program.cs [hunk 13]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+            using (var adminClient = new AdminClientBuilder(new AdminClientConfig { BootstrapServers = bootstrapServers }).Build())
+            {
+                try
+                {
+                    var results = await adminClient.CreateAclsAsync(aclBindings);
+                    Console.WriteLine("All the create ACL operations completed successfully");
+                }
+                catch (CreateAclsException e)
+                {
+                    Console.WriteLine("Some ACL operations failed");
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:190]: maybe "One or more create ACL operations failed."

File: examples/AdminClient/Program.cs [hunk 14]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                {
+                    Console.WriteLine($"An error occurred calling the CreateAcls operation: {e.Message}");
+                }
+            }
+        }
+
+        static async Task DescribeAclsAsync(string bootstrapServers, string[] commandArgs)
+        {
+            var numCommandArgs = commandArgs.Length;
+            var printUsage = numCommandArgs != 7;
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:216]: same comment as before, rearrange to use exceptions + the nargs check taken care of elsewhere.

File: examples/AdminClient/Program.cs [hunk 15]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                Environment.ExitCode = 1;
+                return;
+            }
+
+            using (var adminClient = new AdminClientBuilder(new AdminClientConfig { BootstrapServers = bootstrapServers }).Build())
+            {
+                try
+                {
+                    var result = await adminClient.DescribeAclsAsync(aclBindingFilters[0]);
+                    Console.WriteLine($"Matching ACLs");
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:237]: don't need the `$`. and maybe `"Matching ACLs:"`

File: examples/AdminClient/Program.cs [hunk 16]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                    PrintAclBindings(result.AclBindings);
+                }
+                catch (DescribeAclsException e)
+                {
+                    var result = e.Result;
+                    Console.WriteLine("Describe ACLs operation failed");
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:243]: get rid of this line i think

File: examples/AdminClient/Program.cs [hunk 17]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+            {
+                try
+                {
+                    var result = await adminClient.DescribeAclsAsync(aclBindingFilters[0]);
+                    Console.WriteLine($"Matching ACLs");
+                    PrintAclBindings(result.AclBindings);
+                }
+                catch (DescribeAclsException e)
+                {
+                    var result = e.Result;
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:242]: there's very little reason to define result, just use e.Result below.

File: examples/AdminClient/Program.cs [hunk 18]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                {
+                    Console.WriteLine($"An error occurred calling the describe ACLs operation: {e.Message}");
+                }
+            }
+        }
+
+        static async Task DeleteAclsAsync(string bootstrapServers, string[] commandArgs)
+        {
+            var numCommandArgs = commandArgs.Length;
+            var printUsage = numCommandArgs == 0 || numCommandArgs % 7 != 0;
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:257]: same comment about reorganizing

File: examples/AdminClient/Program.cs [hunk 19]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+            List<AclBindingFilter> aclBindingFilters = null;
+            if (!printUsage)
+            {
+                aclBindingFilters = ParseAclBindingFilters(commandArgs);
+            }
+            printUsage = aclBindingFilters == null;
+
+            if (printUsage)
+            {
+                Console.WriteLine("usage: .. <bootstrapServers> delete-acls <resource_type1> <resource_name1> <resource_patter_type1> "+
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:267]: space before `+`

File: examples/AdminClient/Program.cs [hunk 20]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                return;
+            }
+
+            using (var adminClient = new AdminClientBuilder(new AdminClientConfig { BootstrapServers = bootstrapServers }).Build())
+            {
+                try
+                {
+                    var results = await adminClient.DeleteAclsAsync(aclBindingFilters);
+                    Console.WriteLine("All the delete ACL operations completed successfully");
+                    int i = 0;
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:279]: maybe don't bother printing this out, as the user specified it anyway?

File: examples/AdminClient/Program.cs [hunk 21]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                    foreach (var result in results)
+                    {
+                        Console.WriteLine($"Deleted ACLs in operation {i}");
+                        PrintAclBindings(result.AclBindings);
+                        ++i;
+                    }
+                }
+                catch (DeleteAclsException e)
+                {
+                    Console.WriteLine("Some ACL operations failed");
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:289]: "One or more Delete ACL operation failed"

File: examples/AdminClient/Program.cs [hunk 22]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                    int i = 0;
+                    foreach (var result in e.Results)
+                    {
+                        if (!result.Error.IsError)
+                        {
+                            Console.WriteLine($"Delete ACLs operation {i} completed successfully");
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:295]: i wouldn't bother doing anything in the success case i don't think.

File: examples/AdminClient/Program.cs [hunk 23]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                            Console.WriteLine($"Deleted ACLs in operation {i}");
+                            PrintAclBindings(result.AclBindings);
+                        }
+                        else
+                        {
+                            Console.WriteLine($"An error occurred in delete ACL operation {i}: Code: {result.Error.Code}" +
+                            $", Reason: {result.Error.Reason}");
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:302]: tab indent

File: examples/AdminClient/Program.cs [hunk 24]
```diff
@@ -84,7 +95,218 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                    {
+                        Console.WriteLine($"Deleted ACLs in operation {i}");
+                        PrintAclBindings(result.AclBindings);
+                        ++i;
+                    }
+                }
+                catch (DeleteAclsException e)
+                {
+                    Console.WriteLine("Some ACL operations failed");
+                    int i = 0;
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:290]: since you need the `i`, i think it's better to use a for(var i=0 ... loop.

File: src/Confluent.Kafka/Admin/AccessControlEntry.cs [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+
```
Comment (mhowlett) [inline src/Confluent.Kafka/Admin/AccessControlEntry.cs:17]: by convention, one new line between copyright and using statements :-)

File: src/Confluent.Kafka/Admin/CreateAclResult.cs [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+using System;
+
+
+namespace Confluent.Kafka.Admin
+{
+    /// <summary>
+    ///     Provides create ACL error information.
+    /// </summary>
+    public class CreateAclResult
```
[the remark this answers is not in the document]
Comment (emasab) [reply src/Confluent.Kafka/Admin/CreateAclResult.cs:25]: so the list of CreateAclResult (or better CreateAclReport ) should be only in the exception and the CreateAclsAsync should return void right?

[the remark this answers is not in the document]
Comment (mhowlett) [reply src/Confluent.Kafka/Admin/CreateAclResult.cs:25]: yep (should return Task because it's async)

File: src/Confluent.Kafka/Admin/AbstractAclsResult.cs [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+
+namespace Confluent.Kafka.Admin
+{
+    /// <summary>
+    ///    Base class for ACL results with an <see cref="Error" /> and a list of <see cref="AclBinding" />
+    /// </summary>
+    public abstract class AbstractAclsResult
```
[the remark this answers is not in the document]
Comment (emasab) [reply src/Confluent.Kafka/Admin/AbstractAclsResult.cs:27]: Yes, I wasn't sure of this as it's not clear if they'll diverge in the future. I was thinking that they could be duplicated when needed.

**Comment by emasab** [comment]
I've put for ACLs that the Report aggregates a Result, and an Error. And the Result doesn't have an Error. This way subsequent changes to the result don't affect both classes and it's easier to extract the result from the Report

File: .travis.yml (hunk 1 above)
Comment (mhowlett) [inline .travis.yml:11]: note: i'm moving this to .NET 6 in 1.9.1, required for M1 compatibility.

File: examples/AdminClient/Program.cs [hunk 28]
```diff
@@ -84,7 +95,205 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+        }
+
+        static List<AclBinding> ParseAclBindings(string[] args, bool many)
+        {
+            var numCommandArgs = args.Length;
+            if (
+                many ?
+                (numCommandArgs == 0 || numCommandArgs % 7 != 0) :
+                numCommandArgs != 7
+            )
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:110]: i would format this: 

```
if (many
           ? (numCommandArgs == 0 || numCommandArgs % 7 != 0)
           : numCommandArgs != 7)
{
```

i'd actually change some other things about this but it's being picky and not worth worrying about.

File: examples/AvroGeneric/Program.cs [hunk 29]
```diff
@@ -98,13 +98,13 @@ static async Task Main(string[] args)
             {
                 Console.WriteLine($"{producer.Name} producing on {topicName}. Enter user names, q to exit.");
 
-                long i = 1;
+                long i = 0;
```
Comment (mhowlett) [inline examples/AvroGeneric/Program.cs:101]: future note: generally best to avoid doing unrelated changes - to keep history/blame as useful as possible. but that's fine, just noting for the future.

File: src/Confluent.Kafka/Admin/DeleteAclsReport.cs [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+{
+    /// <summary>
+    ///    Report for a delete ACLs operation with an <see cref="Error" /> and a list of <see cref="AclBinding" />
+    /// </summary>
+    public class DeleteAclsReport
+    {
+        /// <summary>
+        ///     Result of this report
+        /// </summary>
+        public DeleteAclsResult Result { get; set; }
```
Comment (mhowlett) [inline src/Confluent.Kafka/Admin/DeleteAclsReport.cs:32]: I think it's better if you don't use DeleteAclsResult here and replicate the property instead. Regardless, this is the convention set by DeleteRecords, so good to follow it.

File: src/Confluent.Kafka/Admin/DeleteAclsResult.cs [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+{
+    /// <summary>
+    ///    Result for a delete ACLs operation with a list of <see cref="AclBinding" />
+    /// </summary>
+    public class DeleteAclsResult
+    {
+        /// <summary>
+        ///     List of ACL bindings in this result
+        /// </summary>
+        public List<AclBinding> AclBindings { get; set; }
```
Comment (mhowlett) [inline src/Confluent.Kafka/Admin/DeleteAclsResult.cs:31]: In the other admin operations, this is not a list here, rather a list is returned from the DeleteAclsAsync method.
We could argue about which way is better (i think the way the other methods are), but regardless, i think consistency wins out given what we have.

File: examples/AvroGeneric/Program.cs (hunk 29 above)
Comment (emasab) [reply examples/AvroGeneric/Program.cs:101]: Ok

File: src/Confluent.Kafka/Admin/DeleteAclsResult.cs (hunk 31 above)
Comment (mhowlett) [reply src/Confluent.Kafka/Admin/DeleteAclsResult.cs:31]: ahh, i understand now why you want a list here. you want to return a list of lists.

Comment (emasab) [reply src/Confluent.Kafka/Admin/DeleteAclsResult.cs:31]: That is like this because each filter returns a different list of ACL bindings that have been deleted

File: src/Confluent.Kafka/Admin/DeleteAclsReport.cs (hunk 30 above)
Comment (emasab) [reply src/Confluent.Kafka/Admin/DeleteAclsReport.cs]: Done

File: examples/AdminClient/Program.cs (hunk 28 above)
Comment (emasab) [reply examples/AdminClient/Program.cs]: Done

File: .travis.yml (hunk 1 above)
Comment (emasab) [reply .travis.yml:11]: Yeah, good!

File: src/Confluent.Kafka/Admin/DeleteAclsException.cs [hunk 32]
```diff
@@ -0,0 +1,130 @@
...
+        {
+            Reports = reports;
+        }
+
+        /// <summary>
+        ///     The result corresponding to all the delete ACLs operations in the request 
+        ///     (whether or not they were in error). At least one of these
+        ///     results will be in error.
+        /// </summary>
+        public List<DeleteAclsReport> Reports { get; }
```
Comment (mhowlett) [inline src/Confluent.Kafka/Admin/DeleteAclsException.cs:49]: I agree that `Reports` is a better name, but other exception classes use `Results`, so i think change this.

File: src/Confluent.Kafka/Admin/Extensions/String.cs [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+
+
+namespace Confluent.Kafka.Admin
+{
+    /// <summary>
+    ///     Extension methods for the <see cref="String"/> class.
+    /// </summary>
+    internal static class StringExtensions
+    {
+        internal static string Quote(this string str) =>
```
Comment (mhowlett) [inline src/Confluent.Kafka/Admin/Extensions/String.cs:28]: it would be better to put this in here: https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/src/Confluent.Kafka/Internal/Extensions/String.cs

File: src/Confluent.Kafka/AdminClient.cs [hunk 34]
```diff
@@ -123,6 +123,76 @@ private List<DescribeConfigsReport> extractResultConfigs(IntPtr configResourcesP
...
+                }
+            ).ToList();
+        }
+
+        private DescribeAclsReport extractDescribeAclsReport(IntPtr resultPtr)
+        {
+            var errCode = Librdkafka.event_error(resultPtr);
+            var errString = Librdkafka.event_error_string(resultPtr);
+            var resultAcls = Librdkafka.DescribeAcls_result_acls(resultPtr,
+                                                                out UIntPtr resultAclCntPtr);
```
Comment (mhowlett) [inline src/Confluent.Kafka/AdminClient.cs:170]: format like:

```
var resultAcls = Librdkafka.DescribeAcls_result_acls(
                        resultPtr, out UIntPtr resultAclCntPtr);
```

File: src/Confluent.Kafka/AdminClient.cs [hunk 35]
```diff
@@ -123,6 +123,76 @@ private List<DescribeConfigsReport> extractResultConfigs(IntPtr configResourcesP
...
+        private List<DeleteAclsReport> extractDeleteAclsReports(IntPtr resultPtr)
+        {
+            var resultResponsesPtr = Librdkafka.DeleteAcls_result_responses(resultPtr, out UIntPtr resultResponsesCntPtr);
+
+            IntPtr[] resultResponsesPtrArr = new IntPtr[(int)resultResponsesCntPtr];
+            Marshal.Copy(resultResponsesPtr, resultResponsesPtrArr, 0, (int)resultResponsesCntPtr);
+
+            return resultResponsesPtrArr.Select(resultResponsePtr => {
+                var matchingAcls = Librdkafka.DeleteAcls_result_response_matching_acls(resultResponsePtr,
+                                                                                       out UIntPtr resultResponseAclCntPtr);
```
Comment (mhowlett) [inline src/Confluent.Kafka/AdminClient.cs:187]: same comment on formatting as above. put resultResponsePtr on a new line, indent it past the = on the prev line, and out UIntPtr resultResponseAclCntPtr immediately after it.

File: src/Confluent.Kafka/IAdminClient.cs [hunk 36]
```diff
@@ -214,6 +214,103 @@ public interface IAdminClient : IClient
...
+        /// </summary>
+        /// <param name="aclBindings">
+        ///     A IEnumerable with the ACL bindings to create.
+        /// </param>
+        /// <param name="options">
+        ///     The options to use when creating the ACL bindings.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        ///     Thrown if <paramref name="aclBindings"/> param is null
+        ///     or one the <see cref="Confluent.Kafka.Admin.AclBinding.Entry"/> is null or
```
Comment (mhowlett) [inline src/Confluent.Kafka/IAdminClient.cs:229]: "or one the" -> "or a"

File: src/Confluent.Kafka/IAdminClient.cs [hunk 37]
```diff
@@ -214,6 +214,103 @@ public interface IAdminClient : IClient
...
+        ///     one <see cref="Confluent.Kafka.Admin.AclBinding.Pattern"/> is null.
```
Comment (mhowlett) [inline src/Confluent.Kafka/IAdminClient.cs:230]: "or one" -> "or a"

File: test/Confluent.Kafka.IntegrationTests/testconf.json [hunk 38]
```diff
@@ -1,4 +1,4 @@
 {
-  "bootstrapServers": "127.0.0.1:9092",
+  "bootstrapServers": "127.0.0.1:29092",
```
Comment (mhowlett) [inline test/Confluent.Kafka.IntegrationTests/testconf.json:2]: revert this back to 9092

File: test/Confluent.Kafka.UnitTests/Admin/CreateTopicsError.cs [hunk 39]
```diff
@@ -26,7 +26,7 @@ public class CreateTopicsErrorTests
         [Fact]
         public async void NullTopic()
         {
-            using (var adminClient = new AdminClientBuilder(new AdminClientConfig { BootstrapServers = "localhost:666" }).Build())
+            using (var adminClient = new AdminClientBuilder(new AdminClientConfig { BootstrapServers = "localhost:90922" }).Build())
```
Comment (mhowlett) [inline test/Confluent.Kafka.UnitTests/Admin/CreateTopicsError.cs:29]: don't make unnecessary changes

File: test/Confluent.Kafka.UnitTests/Admin/CreateTopicsError.cs [hunk 40]
```diff
@@ -38,7 +38,7 @@ public async void NullTopic()
         [Fact]
         public async void LocalTimeout()
         {
-            using (var adminClient = new AdminClientBuilder(new AdminClientConfig { BootstrapServers = "localhost:666" }).Build())
+            using (var adminClient = new AdminClientBuilder(new AdminClientConfig { BootstrapServers = "localhost:90922" }).Build())
```
Comment (mhowlett) [inline test/Confluent.Kafka.UnitTests/Admin/CreateTopicsError.cs:41]: unnecessary change

File: test/docker/docker-compose.yaml [hunk 41]
```diff
@@ -16,32 +16,52 @@ services:
     image: confluentinc/cp-kafka
     depends_on:
       - zookeeper
+    ports:
+      - 29092:29092
```
Comment (mhowlett) [inline test/docker/docker-compose.yaml:20]: it would be good if you could expose 9092, not 29092, because then you can also test without docker by just running kafka locally, which is how i like to do it.

File: test/Confluent.SchemaRegistry.Serdes.IntegrationTests/testconf.json [hunk 42]
```diff
@@ -1,4 +1,4 @@
 ﻿{
-  "bootstrapServer": "localhost:9092",
+  "bootstrapServer": "localhost:29092",
```
Comment (mhowlett) [inline test/Confluent.SchemaRegistry.Serdes.IntegrationTests/testconf.json:2]: 9092

File: test/Confluent.SchemaRegistry.IntegrationTests/testconf.json [hunk 43]
```diff
@@ -0,0 +1,4 @@
+﻿{
+  "bootstrapServer": "localhost:29092",
```
Comment (mhowlett) [inline test/Confluent.SchemaRegistry.IntegrationTests/testconf.json:2]: 9092

**Review by mhowlett — COMMENTED** [review]
last round of comments... i think. have looked at it all closely now, but it's big, so i'll look through again after these comments addressed before approving.

File: examples/AdminClient/Program.cs [hunk 44]
```diff
@@ -84,7 +95,202 @@ static async Task CreateTopicAsync(string bootstrapServers, string topicName)
...
+                    Console.WriteLine($"An error occurred creating topic {e.Results[0].Topic}: {e.Results[0].Error.Reason}");
+                }
+            }
+        }
+
+        static List<AclBinding> ParseAclBindings(string[] args, bool many)
+        {
+            var numCommandArgs = args.Length;
+            if (many ? (numCommandArgs == 0 || numCommandArgs % 7 != 0)
+                    : numCommandArgs != 7)
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:107]: one more space ...

File: examples/AdminClient/Program.cs [hunk 45]
```diff
@@ -121,9 +121,18 @@ static List<AclBinding> ParseAclBindings(string[] args, bool many)
                 var operation = Enum.Parse<AclOperation>(args[baseArg + 5]);
                 var permissionType = Enum.Parse<AclPermissionType>(args[baseArg + 6]);
 
-                if (name == "") name = null;
-                if (principal == "") principal = null;
-                if (host == "") host = null;
+                if (name == "")
+                {
```
Comment (mhowlett) [inline examples/AdminClient/Program.cs:125]: this is fine, but i would usually put it on one line
`if (name == "") { name = null; }`
